crates/noirc_frontend/src/hir/def_collector/errors.rs

```diff
@@ -30,60 +36,55 @@ impl DefCollectorErrorKind {
     }
 }
 
+fn report_duplicate(
+    primary_message: String,
+    duplicate_type: &str,
+    first_def: Ident,
+    second_def: Ident,
+) -> Diagnostic {
+    let first_span = first_def.0.span();
+    let second_span = second_def.0.span();
+    let mut diag = Diagnostic::simple_error(
+        primary_message,
+        format!("first {} found here", duplicate_type),
+        first_span,
+    );
+    diag.add_secondary(format!("second {} found here", duplicate_type), second_span);
+    diag
+}
+
 impl From<DefCollectorErrorKind> for Diagnostic {
     fn from(error: DefCollectorErrorKind) -> Diagnostic {
         match error {
             DefCollectorErrorKind::DuplicateFunction { first_def, second_def } => {
-                let first_span = first_def.0.span();
-                let second_span = second_def.0.span();
-                let func_name = &first_def.0.contents;
-
-                let mut diag = Diagnostic::simple_error(
-                    format!("duplicate definitions of {func_name} function found"),
-                    "first definition found here".to_string(),
-                    first_span,
-                );
-                diag.add_secondary("second definition found here".to_string(), second_span);
-                diag
+                let primary_message =
+                    format!("duplicate definitions of {} function found", &first_def.0.contents);
+                report_duplicate(primary_message, "function definition", first_def, second_def)
+            }
+            DefCollectorErrorKind::DuplicateTypeDef { first_def, second_def } => {
+                let primary_message =
+                    format!("duplicate definitions of {} type found", &first_def.0.contents);
+                report_duplicate(primary_message, "type definition", first_def, second_def)
+            }
+            DefCollectorErrorKind::DuplicateTraitDef { first_def, second_def } => {
+                let primary_message =
+                    format!("duplicate definitions of {} trait found", &first_def.0.contents);
+                report_duplicate(primary_message, "trait definition", first_def, second_def)
```

Comment:
I don't think the one word difference here is useful enough to have 6 separate error kinds for duplicates. At most we should have 1 where the definition kind is a field

Reply:
I agree, I will refactor this.